onboard: 1.4.1

johnramsden opened the pull request:
###### Motivation for this change

[onboard](https://launchpad.net/onboard) is a great on-screen keyboard and in my opinion should be in the repo. 

Everything is working and there are no errors at runtime.

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

FRidh commented:
We have the function `python3.pkgs.buildPythonApplication` for building Python packages. See the manual.

johnramsden commented:
@FRidh I'll take another look. I guess I was incorrect when I thought the python specific functions were just for pip packages.

---

Another thing, when following the manual it says in [the quick start section](https://nixos.org/nixpkgs/manual/#chap-quick-start)

> To test whether the package builds, run the following command from the root of the nixpkgs source tree: `$ nix-build -A libfoo`

When trying that on NixOS after cloning the repo I get 

> error: getting status of ‘/home/john/Workspace/Packaging/NixOS/nixpkgs/pkgs/onboard’: No such file or directory

and I have to run it from the package's directory to get it to work. What am I doing wrong?

FRidh commented:
Write down exactly what you do, since the error message you give does not correspond to the code you show here.

johnramsden commented:
After cloning the repo and setting up my package:

[Step 5](https://nixos.org/nixpkgs/manual/#chap-quick-start):

run the following command from the root of the nixpkgs source tree:

$ nix-build -A  -K onboard

> error: getting status of ‘/home/john/Workspace/Packaging/NixOS/nixpkgs/onboard’: No such file or directory

Mic92 commented on pkgs/applications/misc/onboard/default.nix:
```diff
@@ -0,0 +1,45 @@
+{ gtk3, python3, hunspell, isocodes, libcanberra_gtk3, xorg.libxkbfile, libxkbcommon, python35Packages.pycairo, python35Packages.dbus-python, python35Packages.pygobject3, python35Packages.systemd, libudev, python35Packages.distutils_extra, gnome3.dconf, pkgconfig, xorg.libXtst
```
these are syntax errors.

johnramsden replied:
That's strange, I fixed that in the previous commit, I guess I undid the change by accident.

johnramsden commented:
I tried to convert to use `buildPythonApplication`, I keep getting "error: cannot auto-call a function that has an argument without a default value (‘stdenv’)" when I try to build though...

johnramsden commented:
So i've tested building and it seems to work until an issue with it being unable to find symlinked files, I think it's possibly related to [issue 1232](https://github.com/NixOS/nix/issues/1232). 

> error: [Errno 2] No such file or directory: 'Onboard/pypredict/attic/test-client.py'

It occurs in the [setup.py](http://bazaar.launchpad.net/~onboard/onboard/trunk/view/head:/setup.py), in this code:

``` python
def symlink_extension_libraries(setup_command):
    """
    Link the extensions back to the project directory
    so Onboard can be run from source as usual, without --inplace.
    Remove this at any time if there is a better way.
    """
    if setup_command in ["build", "build_ext"]:
        for path, pattern in libs_to_symlink:
            files = glob.glob(join(build_root, path, pattern))
            for file in files:
                dstfile = join(path, split(file)[1])
                print("symlinking {} to {}".format(file, dstfile))

                try: os.unlink(dstfile)
                except OSError: pass
                os.symlink(file, dstfile)

# Make xgettext extract translatable strings from _format() calls too.
var = "XGETTEXT_ARGS"
os.environ[var] = os.environ.get(var, "") + " --keyword=_format"

# scan for translatable layout strings in layouts
# disabled until know how to make it work.
# layoutstring.px has those string manually added now.
if 0:
    if "build_i18n" in sys.argv:
        args = ["./tools/gen_i18n_strings",
                "-o./data/layoutstrings_generated.py"]
        print("Running '{}'".format(" ".join(args)))
        subprocess.check_call(args)

clean_before_build(setup_command)

```
Any advice?

johnramsden commented:
I opened [a question](https://answers.launchpad.net/onboard/+question/625302) on onboards launchpad page in case they could help.

johnramsden commented:
So, [I was told](https://answers.launchpad.net/onboard/+question/625302) the 'attic directory wasn't neccesary anyway and removed it in the preBuild.

``` shell
    preBuild = ''
      rm -r Onboard/pypredict/attic
      sed -i 's:/bin/bash:${bash}/bin/bash:' ./setup.py
    '';
```
> [marmuta (marmuta)](https://launchpad.net/~marmuta) said 1 hour ago:	1
> This is probably DistUtilsExtra trying to make sense of the attic files and failing. I'm not sure what exactly happens, you would have to look at auto.py for that.
> However the easy fix is to simply delete the offending file(s). Nothing in the attic folder is even used. We should probably remove the whole folder, just to be safe.
> 
> [marmuta (marmuta)](https://launchpad.net/~marmuta) said 45 minutes ago:	2
> OK, Onboard/pypredict/attic is gone in trunk. The next release won't have it anymore.
> 

Now i'm getting:

```
running build_scripts
creating build/scripts-3.5
copying and adjusting onboard -> build/scripts-3.5
copying and adjusting onboard-settings -> build/scripts-3.5
changing mode of build/scripts-3.5/onboard from 644 to 755
changing mode of build/scripts-3.5/onboard-settings from 644 to 755
running build_i18n
intltool-update -p -g onboard
error: [Errno 2] No such file or directory: 'Onboard/pypredict/tools/model_info.py'
```

johnramsden commented:
So the [onboard guys](https://answers.launchpad.net/onboard/+question/625302) gave me some input and I was able to get onboard building, I'm getting some  errors though when run. I'm not sure if it's a Nix thing or an onboard thing. 

I also changed things to use the Nix specific functions. 

I'm now getting the following error when onboard is run: 

```
Traceback (most recent call last):
  File "/nix/store/ar8sjdlrwvaq30gp4mgc3y4xz8nwsy7x-onboard-1.4.1/bin/..onboard-wrapped-wrapped", line 36, in <module>
    from Onboard.OnboardGtk import OnboardGtk as Onboard
  File "/nix/store/ar8sjdlrwvaq30gp4mgc3y4xz8nwsy7x-onboard-1.4.1/lib/python3.5/site-packages/Onboard/OnboardGtk.py", line 35, in <module>
    require_gi_versions()
  File "/nix/store/ar8sjdlrwvaq30gp4mgc3y4xz8nwsy7x-onboard-1.4.1/lib/python3.5/site-packages/Onboard/Version.py", line 24, in require_gi_versions
    gi.require_version('Gtk', '3.0')
  File "/nix/store/sgqhypg37ad4gvgsj5r9gvz1a680xzh3-python3.5-pygobject-3.22.0/lib/python3.5/site-packages/gi/__init__.py", line 118, in require_version
    raise ValueError('Namespace %s not available' % namespace)
ValueError: Namespace Gtk not available 

```

johnramsden commented:
It looks like some typelibs are missing.
> [marmuta (marmuta) said 16 hours ago:](https://answers.launchpad.net/onboard/+question/625302)
> Great, getting closer.
> Is Gtk-3.0.typelib installed? On Ubuntu this is a separate package gir1.2-gtk-3.0 that contains
> /usr/lib/x86_64-linux-gnu/girepository-1.0/Gdk-3.0.typelib
> /usr/lib/x86_64-linux-gnu/girepository-1.0/GdkX11-3.0.typelib
> /usr/lib/x86_64-linux-gnu/girepository-1.0/Gtk-3.0.typelib
> 
> Onboard uses all of these and a couple other typelibs:
> grep gir debian/control
> Depends: gir1.2-gdkpixbuf-2.0,
>          gir1.2-glib-2.0,
>          gir1.2-gtk-3.0,
>          gir1.2-pango-1.0 (>= 1.29.3),
> Recommends: gir1.2-appindicator3-0.1,
>             gir1.2-atspi-2.0,
> 
> appindicator3 is for the status icon - optional if the desktop environment is fine with the old GtkStatusIcon. I believe KDE Plasma needs it.
> Atspi is for auto-show, word suggestions and others, but not strictly required for a plain keyboard.

I have found the typelibs in `<nix store>/lib/girepository-1.0`
I can't seem to find a package that provides this but I [have seen](https://github.com/peti/nixpkgs/blob/61977a38216021edd814e34ca1da6874afb59622/pkgs/applications/misc/redshift/default.nix) other people set a variable `$GI_TYPELIB_PATH` with wrapProgram. 

I don't want to just blindly do something if I don't understand why i'm doing it. Is this the proper way to set the typelibs?

johnramsden commented:
@FRidh I tried wrapping the package but it didn't help. 

```
 wrapProgram "$out/usr/bin/nylas-mail" \
      --prefix PYTHONPATH : "$PYTHONPATH" \
      --prefix GI_TYPELIB_PATH : "$GI_TYPELIB_PATH"
```

Is there a better way to set the typelibs?

johnramsden commented:
Should I repost this as an issue?_

johnramsden commented:
Okay, I finally made some progress. Adding the packages gobjectIntrospection gsettings_desktop_schemas, and wrapGAppsHook it seems to boot but now I'm getting a different error:

```
** (.onboard-wrapped:6052): WARNING **: Error retrieving accessibility bus address: org.freedesktop.DBus.Error.ServiceUnknown: The name org.a11y.Bus was not provided by any .service files
17:05:26.646 WARNING Onboard.Keyboard: Atspi typelib missing, at-spi key-synth unavailable
17:05:26.656 WARNING Onboard.AtspiStateTracker: Atspi typelib missing, auto-show unavailable
17:05:26.682 WARNING Config: mousetweaks GSettings schema not found, mousetweaks integration disabled.
Traceback (most recent call last):
  File "/nix/store/l8bry5j9nf10kk0vgchx9ghmpblr4qhx-onboard-1.4.1/bin/..onboard-wrapped-wrapped", line 37, in <module>
    ob = Onboard()
  File "/nix/store/l8bry5j9nf10kk0vgchx9ghmpblr4qhx-onboard-1.4.1/lib/python3.5/site-packages/Onboard/OnboardGtk.py", line 148, in __init__
    self.init()
  File "/nix/store/l8bry5j9nf10kk0vgchx9ghmpblr4qhx-onboard-1.4.1/lib/python3.5/site-packages/Onboard/OnboardGtk.py", line 178, in init
    config.init()
  File "/nix/store/l8bry5j9nf10kk0vgchx9ghmpblr4qhx-onboard-1.4.1/lib/python3.5/site-packages/Onboard/Config.py", line 370, in init
    self.install_dir = self._get_install_dir()
  File "/nix/store/l8bry5j9nf10kk0vgchx9ghmpblr4qhx-onboard-1.4.1/lib/python3.5/site-packages/Onboard/Config.py", line 1449, in _get_install_dir
    assert(result) # warn early when the installation dir wasn't found
AssertionError
```

johnramsden commented:
Whoo! Got it running. Just need to fix a few errors and it should be ready!

johnramsden commented:
The issue with seems to be related to how the file sokSettings.py is being wrapped. I'm using wrapGAppsHook and the program wants to import a module but it is importing the bash script wrapper which actually calls the module, instead of importing the Python module. Does anyone know if there is a way to exclude certain files with the wrappers?

---

When the Python script attempts to import sokSettings.py the following happens:

```
python3 -c "import scripts.sokSettings; scripts.sokSettings.run()"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/nix/store/5cwnb0da5rjm9ni4pf1fg7i2ng6k5c8j-onboard-1.4.1/share/onboard/scripts/sokSettings.py", line 2
    export PATH=/nix/store/76rgpkm1jqgqwz4bjnk8y37z29q82zgy-python3-3.5.3/bin:/nix/store/5cwnb0da5rjm9ni4pf1fg7i2ng6k5c8j-onboard-1.4.1/bin:/nix/store/63spjx86svppz695s5yy7mh2h6lbbca1-gtk+3-3.22.8-dev/bin:/nix/store/k6bmqyh6yf8pzrhnfbzkqdrb3g0hb5gj-expat-2.2.0-dev/bin:/nix/store/pmp2ysvnrmcfzdjmaqnyhda96zkf2z6f-glib-2.50.3-dev/bin:/nix/store/8xm32qg4xi1l41zng4xv752492h6wmdp-cairo-1.14.8-dev/bin:/nix/store/dg3c94l36n3zmrpq9p9qyfi6407rhv5f-freetype-2.6.5-dev/bin:/nix/store/6fc4ak60p1mnilrjfcyz0sywwmpmf45w-bzip2-1.0.6.0.1-bin/bin:/nix/store/j8japwvk2qq8wf9yddibs4fnp99hskkw-libpng-apng-1.6.28-dev/bin:/nix/store/fsqw38ydj8wfj606rzkdcb3rrmpnhb4q-fontconfig-2.12.1-bin/bin:/nix/store/3hz3ll4ikznx3b6cd3myvrm5i79kyj8y-harfbuzz-1.4.2-dev/bin:/nix/store/pr6x655sz1n1sqjhch2i64h5l8i48xrm-graphite2-1.3.6/bin:/nix/store/7wc7qlqg7vsbav3vrjmkfbjbxpp9nnpr-pango-1.40.3-bin/bin:/nix/store/mrzs48prbr7aync14ns4l2j2s489sx50-gdk-pixbuf-2.36.5-dev/bin:/nix/store/sqxw6zlqy91y5nhl4smr5smixkqhdr1k-libjpeg-turbo-1.5.1-bin/bin:/nix/store/s2hmf51bprsxb20v3bfigs8a7zsw3hjl-xz-5.2.2-bin/bin:/nix/store/i14lfjn7cgfmghak3hqczmd90dymsbaj-libtiff-4.0.7-bin/bin:/nix/store/8iavg1dpbz7w9lqgv35q0a6hdkr0gpxi-jasper-2.0.12-bin/bin:/nix/store/pj75dhcgcnnv7ws860bs2zf23hzk2bjz-gdk-pixbuf-2.36.5/bin:/nix/store/qja3mjgiz9c7ld1ddffgc7b487mv1zmh-gobject-introspection-1.50.0-dev/bin:/nix/store/6l6kkkvbz9nxc7pi8m4wcq0z60bmp74a-wayland-1.12.0/bin:/nix/store/q5l53jgm2crfih93l56lcl8mxjywygs9-cups-2.1.4-dev/bin:/nix/store/3gi4zw66q73hdal18akrca30rf02zblq-cups-2.1.4/bin:/nix/store/bx1091kbmwff89gmmxy3za426ryxg0jm-gtk+3-3.22.8/bin:/nix/store/2byhribnzhiy7zhp838fc833xdhnl5l8-dconf-0.26.0/bin:/nix/store/iwm7yk3a40wn400wlqv0rx0zhc7rk756-libcanberra-0.30/bin:/nix/store/px7a1kq4prjzw17c9pppssdmibqqdm2p-systemd-232/bin:/nix/store/fi3mbd2ml4pbgzyasrlnp0wyy6qi48fh-bash-4.4-p5/bin:/nix/store/nqini3ad62zhsknqdpsjcgcm68d1n4gl-hunspell-1.3.3-bin/bin:/nix/store/q2djjbafrmarpv3iqc6scqnsxrdlqxvb-pkg-config-0.29/bin:/nix/store/0l73r9z7q5zjyvghnkd9wm8bihf9gg2p-intltool-0.51.0/bin:/nix/store/lkdr3z34n6n17g7xiaw8plyw7hv9idsl-gettext-0.19.8/bin:/nix/store/nsa311yg8h93wfaacjk16c96a98bs09f-perl-5.22.3/bin:/nix/store/4yn2vyzz1mwl9ry9l67y9bmyasjp7jrj-python3.5-setuptools-30.2.0/bin:/nix/store/087w6yxwq3skp2zppahq8dvfa2idckv7-librsvg-2.40.16/bin${PATH:+:}$PATH
              ^
SyntaxError: invalid syntax
```

This is what it should be importing:

```
cat share/onboard/scripts/.sokSettings.py-wrapped
#!/nix/store/76rgpkm1jqgqwz4bjnk8y37z29q82zgy-python3-3.5.3/bin/python3

import sys;import site;import functools;sys.argv[0] = '/nix/store/5cwnb0da5rjm9ni4pf1fg7i2ng6k5c8j-onboard-1.4.1/share/onboard/scripts/sokSettings.py';functools.reduce(lambda k, p: site.addsitedir(p, k), ['/nix/store/5cwnb0da5rjm9ni4pf1fg7i2ng6k5c8j-onboard-1.4.1/lib/python3.5/site-packages','/nix/store/fs8cxcjdm2zf8arxmbva7k742n7mkklm-python3.5-pycairo-1.10.0/lib/python3.5/site-packages','/nix/store/4yn2vyzz1mwl9ry9l67y9bmyasjp7jrj-python3.5-setuptools-30.2.0/lib/python3.5/site-packages','/nix/store/7kfnj63b4pav99zgq24cjxx3cjgl3xq2-python3.5-dbus-python-1.2.4/lib/python3.5/site-packages','/nix/store/168n92ld073flhld87hz4nv39vr44rd7-python3.5-pygobject-3.22.0/lib/python3.5/site-packages','/nix/store/3j9whdk7zqif6jj9alrm8i3d0fx8ma7c-python3.5-python-systemd-231/lib/python3.5/site-packages','/nix/store/42dan8qv7ffjbwzqcnzdqw3mr16imwzp-python3.5-distutils-extra-2.39/lib/python3.5/site-packages','/nix/store/6f4piaq9la15ljqjjz0fyzf653kcfj5l-pyatspi-2.18.0/lib/python3.5/site-packages'], site._init_pathinfo());
PYTHON_EXECUTABLE = "python3"

from gi.repository import GLib

def run():
    GLib.spawn_async(argv=[PYTHON_EXECUTABLE,
                           "-cfrom Onboard.settings import Settings\ns = Settings(False)"],
                     flags=GLib.SpawnFlags.SEARCH_PATH)
```

johnramsden commented:
Now the setting application actually runs.

I did hardcoded it to execute the specific executable of the settings module, how it was set before stopped settings from starting.

```
      substituteInPlace ./scripts/sokSettings.py \
        --replace "#!/usr/bin/python3" "" \
        --replace "PYTHON_EXECUTABLE," "\"$out/bin/onboard-settings\"" \
        --replace '"-cfrom Onboard.settings import Settings\ns = Settings(False)"' ""
```

This ends up giving us the following:

```
def run():
    GLib.spawn_async(argv=["/nix/store/7qwvy80hj1spg513944lrspwjl2nsrn9-onboard-1.4.1/bin/onboard-settings"
                           ],
                     flags=GLib.SpawnFlags.SEARCH_PATH)
```

---

The issue is the settings application doesn't do anything. It won't change any settings. I'm wondering if it could be related to the fact that the store is considered read only?

Mic92 commented on pkgs/applications/misc/onboard/default.nix:
```diff
@@ -0,0 +1,122 @@
+        --replace "/usr/lib" "$out/lib"
+
+      substituteInPlace  ./data/org.onboard.Onboard.service  \
+        --replace "/usr/bin" "out/bin"
```
you are missing '$' in $out here and below.

johnramsden replied:
Thanks.

johnramsden commented:
So, marmuta from onboard thinks it might have something to do with 'gsettings'. 

> [marmuta (marmuta) said 14 hours ago:	#36](https://answers.launchpad.net/onboard/+question/625302#yui_3_10_3_1_1497737303187_276)
> Do things like
> gsettings get org.onboard layout
> work?
> Or try running dconf-editor for the same with a GUI tool.

I tried adding a few packages that I thought might might provide gsettings as build inputs, but it doesn't seem to do anything or add the package. 
```
buildInputs = [ gnome3.gconf gnome3.dconf gnome3.dconf-editor ];
```

I'm not sure what should be providing gsettings.

johnramsden commented:
For some reason glib schemas are not being installed. I posted an [issue](https://github.com/NixOS/nixpkgs/issues/26702) regarding the problem.

johnramsden commented:
It looks like the problem I was getting with the settings was not related to the glib schema compilation. I found out that the fix was to install `gnome3.dconf` system-wide. For some reason just adding it to `buildInputs` doesn't work. 

I reverted the schema back to how it was before, which seems to be working. 

The on-screen keyboard is now in a usable state. I'm now working on getting rid of the remaining warnings that show up at the start of the program: 

``` 
** (.onboard-wrapped:16998): WARNING **: Error retrieving accessibility bus address: org.freedesktop.DBus.Error.ServiceUnknown: The name org.a11y.Bus was not provided by any .service files
14:40:03.814 WARNING Onboard.Keyboard: Atspi typelib missing, at-spi key-synth unavailable
14:40:03.824 WARNING Onboard.AtspiStateTracker: Atspi typelib missing, auto-show unavailable
14:40:03.851 WARNING Config: mousetweaks GSettings schema not found, mousetweaks integration disabled.

(onboard:16998): Gtk-WARNING **: Theme parsing error: gtk.css:68:35: The style property GtkButton:child-displacement-x is deprecated and shouldn't be used anymore. It will be removed in a future version

(onboard:16998): Gtk-WARNING **: Theme parsing error: gtk.css:69:35: The style property GtkButton:child-displacement-y is deprecated and shouldn't be used anymore. It will be removed in a future version

(onboard:16998): Gtk-WARNING **: Theme parsing error: gtk.css:73:46: The style property GtkScrolledWindow:scrollbars-within-bevel is deprecated and shouldn't be used anymore. It will be removed in a future version
```

johnramsden commented:
Got rid of last few warnings. Added at_spi2_core as optional, enabling atspiSupport gets rid of warning:

```
17:16:03.997 WARNING Onboard.Keyboard: Atspi typelib missing, at-spi key-synth unavailable
17:16:04.009 WARNING Onboard.AtspiStateTracker: Atspi typelib missing, auto-show unavailable
```

Enable service 'services.gnome3.at-spi2-core.enable = true' to get rid of warning:

```
WARNING **: Error retrieving accessibility bus address: org.freedesktop.DBus.Error.ServiceUnknown: The name org.a11y.Bus was not provided by any .service files
```

grahamc commented on pkgs/top-level/all-packages.nix:
```diff
@@ -1022,7 +1022,10 @@ with pkgs;
 
   mpdris2 = callPackage ../tools/audio/mpdris2 { };
 
+
```
Can you remove this whitespace?

grahamc commented:
Trying to merge this, it seems to not  cleanly apply. Can you rebase to master? Also, please squash the commits in to:


1. adding yourself as a maintainer
2. adding the package with the subject "onboard: init at  1.4.1'


if you need help, I can do this for you.

Graham

johnramsden committed:
onboard: init at 1.4.1

Add testing reqs, but keep tests disabled

Tests are runnable but still produce errors.

To get tests working, add locale setting, replace killall
and add nose package. To run the tests enable 'doCheck'.

Hunspell needs to be explicitly installed to use.

Patch SpellCheck.py to put hunspell in system datadir location.
For example, '/var/run/current-system/sw/share/hunspell/'
or '${HOME}/.nix-profile/share/hunspell/'

To get rid of atspi errors set
'services.gnome3.at-spi2-core.enable = true'

johnramsden commented:
@grahamc All done, let me know if there's anything else.

FRidh commented on pkgs/applications/misc/onboard/default.nix:
```diff
@@ -0,0 +1,134 @@
+{ config
+, pkgs
```
don't pass in `pkgs`, instead pass in individual packages

johnramsden replied:
Right, I think this was leftover from when I was using it in my personal config and I forgot to change it. Same with disabling tests, i'll remove both.

FRidh commented on pkgs/applications/misc/onboard/default.nix:
```diff
@@ -0,0 +1,134 @@
+, xorg
+, libxkbcommon
+, python3
+, python35Packages
```
This is wrong. Either pass in `pythonX` and use `pythonX.pkgs` or pass in `pythonXPackages`. If you pass in both you have the risk they're not the same set. This is exactly what happens here, because `python3` is an alias `python36`.

johnramsden replied:
Is one or the other preferred?

FRidh replied:
Originally `pythonXPackages` was used. Later I introduced `pythonX.pkgs`. I prefer latter but both are fine.

FRidh commented on pkgs/applications/misc/onboard/default.nix:
```diff
@@ -0,0 +1,134 @@
+    sha256 = "01cae1ac5b1ef1ab985bd2d2d79ded6fc99ee04b1535cc1bb191e43a231a3865";
+  };
+
+  doCheck = false;
```
Why do you disable tests? Please add a reason as a comment.

johnramsden replied:
I meant to remove that.

johnramsden replied:
@FRidh So, I tried enabling tests and I'm getting the following error:

```
running install tests
found dconf version 0.26.0
Traceback (most recent call last):
  File "nix_run_setup.py", line 8, in <module>
    exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\\r\\n', '\\n'), __file__, 'exec'))
  File "setup.py", line 440, in <module>
    'build_ext': build_ext_custom},
  File "/nix/store/996yw9bc36jbwy0gn7jpaxvji8i0ijpx-python3.5-distutils-extra-2.39/lib/python3.5/site-packages/DistUtilsExtra/auto.py", line 88, in setup
    __dbus(attrs, src)
  File "/nix/store/996yw9bc36jbwy0gn7jpaxvji8i0ijpx-python3.5-distutils-extra-2.39/lib/python3.5/site-packages/DistUtilsExtra/auto.py", line 192, in __dbus
    lines = [l.strip() for l in open(f).readlines()]
  File "/nix/store/r7qpc32yr09l9a0d5y3b8i84kw5phx4p-python3-3.5.3/lib/python3.5/encodings/ascii.py", line 26, in decode
    return codecs.ascii_decode(input, self.errors)[0]
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc2 in position 12: ordinal not in range(128)
builder for ‘/nix/store/grmh43c2vbw00cvpgrw4ahdgyk6jdxgn-onboard-1.4.1.drv’ failed with exit code 1
cannot build derivation ‘/nix/store/abq3j75wbmgfpjvc5z0nb9y213hjdcwl-system-path.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/i1v54bpqcbigfqp5ab9npqmxvr25gf0y-nixos-system-atom-17.03.git.68b6c50b12M.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/i1v54bpqcbigfqp5ab9npqmxvr25gf0y-nixos-system-atom-17.03.git.68b6c50b12M.drv’ failed
```

It appears to be something [related to unicode](https://stackoverflow.com/questions/21129020/how-to-fix-unicodedecodeerror-ascii-codec-cant-decode-byte). Do you think this is related to the package or could it be related to NixOS's testing?

FRidh replied:
This is a common issue on Nix with Python 3. You need `LC_ALL = "en_US.UTF-8";` and `buildInputs = [ glibcLocales ];`. See `python-packages.nix` for examples.

FRidh commented on pkgs/applications/misc/onboard/default.nix:
```diff
@@ -0,0 +1,134 @@
+
+  doCheck = false;
+
+  propagatedBuildInputs = [
```
I think what you want is to have some libraries available at runtime. Adding them to `propagatedBuildInputs` is not doing that. Instead, what it does is add them to the `buildInputs` of any package that depends on this package.

If your package needs certain libraries during runtime, then it needs to patch the `rpath` or otherwise wrap the executables with a wrapper that sets `LD_LIBRARY_PATH`.

jtojnar replied:
Only the python dependencies should be part of `propagatedBuildInputs`, everything else should go to `buildInputs` if it is a library we want to link against or something we want in any way to be used by the built program, `nativeBuildInputs` if it is only used during building the package (intltool wrapGAppsHook glibcLocales).

johnramsden commented on pkgs/applications/misc/onboard/default.nix:
```diff
@@ -0,0 +1,134 @@
+
+  doCheck = false;
+
+  propagatedBuildInputs = [
```
I'm able to get it to build if I put the inputs I had in `propagatedBuildInputs` in `buildInputs`, I then tried wrapping the program with the following:

```
    wrapProgram $out/bin/onboard \
        --prefix "LD_LIBRARY_PATH" : "${stdenv.lib.makeLibraryPath [
                                                        gtk3
                                                        libcanberra_gtk3
                                                        libudev
                                                        bash
                                                        hunspell
                                                        isocodes
                                                        pkgconfig
                                                        xorg.libXtst
                                                        xorg.libxkbfile
                                                        libxkbcommon
                                                        intltool
                                                        python3.pkgs.pycairo
                                                        python3.pkgs.dbus-python
                                                        python3.pkgs.pygobject3
                                                        python3.pkgs.systemd
                                                        python3.pkgs.distutils_extra
                                                        python3.pkgs.pyatspi
                                                        glib ]}"
```

Is that what you meant? When I use the wrapper I get an error saying the argument list is too long, I get the message even when I'm only using a single library in `stdenv.lib.makeLibraryPath`:

```
/nix/store/7qr5cvy0wrgdqw8c21hfd7clbphm0388-onboard-1.4.1/bin/.onboard-wrapped: line 3: /nix/store/7qr5cvy0wrgdqw8c21hfd7clbphm0388-onboard-1.4.1/bin/.onboard-wrapped: Argument list too long
```
Any thoughts?

FRidh replied:
The Python libraries should be in `propagatedBuildInputs`. Executables that you need during runtime need to be added to `PATH`. Libraries you need during runtime need to be added to `LD_LIBRARY_PATH`, or preferably hardcoded in your package.

johnramsden replied:
@FRidh I'm having trouble figuring out what I need to do here, everything I have tried causes the package to not work. If I use propagatedBuildInputs the program runs fine. Do you have an example of what I should be doing, I can't get this working.

jtojnar commented on pkgs/applications/misc/onboard/default.nix:
```diff
@@ -0,0 +1,133 @@
+  ];
+
+  buildInputs = [
+    glib gobjectIntrospection gsettings_desktop_schemas gnome3.dconf wrapGAppsHook
```
It does not seem `gobjectIntrospection` is required.

jtojnar commented on pkgs/applications/misc/onboard/default.nix:
```diff
@@ -0,0 +1,133 @@
+  ] ++ stdenv.lib.optional atspiSupport at_spi2_core;
+
+  preBuild = ''
+    rm -r Onboard/pypredict/attic
```
Could you add comment why this is necessary?

jtojnar replied:
I see now. It works fine for me with the directory not removed but given the reasoning above, removing it is fine. Link to https://github.com/NixOS/nixpkgs/pull/24986#issuecomment-296114062 would be enough.

jtojnar commented on pkgs/applications/misc/onboard/default.nix:
```diff
@@ -0,0 +1,133 @@
+    rm -r Onboard/pypredict/attic
+
+    substituteInPlace  ./scripts/sokSettings.py \
+    --replace "#!/usr/bin/python3" "" \
```
The shebang can be fixed by `patchShebangs`

jtojnar commented on pkgs/applications/misc/onboard/default.nix:
```diff
@@ -0,0 +1,133 @@
+
+    for dir in ". ./Onboard/ ./scripts/ ./Onboard/test/ ./Onboard/pypredict/
+    ./Onboard/pypredict/lm/ ./Onboard/pypredict/test/ ./Onboard/pypredict/tools/";
+    do patchShebangs $dir; done
```
`patchShebangs` works recursively, `.` is enough.

jtojnar commented on pkgs/applications/misc/onboard/default.nix:
```diff
@@ -0,0 +1,133 @@
+
+    ${glib.dev}/bin/glib-compile-schemas $out/share/glib-2.0/schemas/
+
+    addToSearchPath GI_TYPELIB_PATH $out/lib/girepository-1.0
```
This would be useful if this package provided gir files but then it would be already done by `gobjectIntrospection` hook.

johnramsden replied:
Removing ```glib-compile-schemas``` results in the program erroring out with the following

> ERROR Config: gsettings schema for 'org.onboard.keyboard' not installed

Do you have any suggestions regarding this?

jtojnar replied:
I meant removing the `addToSearchPath`, `glib-compile-schemas` should remain.

jtojnar commented on pkgs/applications/misc/onboard/default.nix:
```diff
@@ -0,0 +1,133 @@
+  '';
+
+  postInstall = ''
+    mkdir -p $out/share/glib-2.0/schemas/ $out/lib/girepository-1.0
```
Creating these directories is pointless. The first one is already created by installation script and the second would be too, if onboard generated gir bindings.

jtojnar commented on pkgs/applications/misc/onboard/default.nix:
```diff
@@ -0,0 +1,133 @@
+    ${glib.dev}/bin/glib-compile-schemas $out/share/glib-2.0/schemas/
+
+    addToSearchPath GI_TYPELIB_PATH $out/lib/girepository-1.0
+    addToSearchPath XDG_DATA_DIRS $out/share
```
Similarly here, `addToSearch` adds to path of packages that use this dependency. And I think it would only work as a part of setup hook anyway.

jtojnar commented on pkgs/applications/misc/onboard/default.nix:
```diff
@@ -0,0 +1,133 @@
+    substituteInPlace  ./setup.py \
+    --replace "/bin/bash" ${stdenv.shell}
+    substituteInPlace  ./Onboard/TextDomain.py \
+    --replace "/bin/bash" "${stdenv.shell}"
```
Also should be handled by `patchShebangs`.

johnramsden commented:
@jtojnar Thanks for the feedback. I don't currently have a desktop system running NixOS, as soon as I can spin up a graphical VM I'll add the changes and make sure it's working.

johnramsden commented:
I'm not sure why but settings only get saved if ```pkgs.gnome3.dconf``` is explicitly installed on the system. Even when I add the package to ```buildInputs```, the settings aren't saved. What would be the reason behind this and the solution?

jtojnar commented:
dconf infrastructure consists of two parts – dconf the library and dconf daemon. In order for the daemon to run, it needs to be installed to system path, inputs only make it available to the package. Using `propagatedUserEnvPkgs` until we fix #11239 might be acceptable.

johnramsden commented:
@jtojnar I added your suggested changes and rebased off master. Let me know if I missed anything.

johnramsden commented:
I tried to get the tests working, and while I got them to run, they do not run successfully.

To get the tests working I replaced ```killall``` with ```pkill -x```, added a locale setting 'LC_ALL = "en_US.UTF-8"', and added 'python3.pkgs.nose'. The tests are now runnable but produce a mass of errors so I left them them disabled for now.

jtojnar commented on pkgs/applications/misc/onboard/default.nix:
```diff
@@ -0,0 +1,144 @@
+  ];
+
+  buildInputs = [
+    #python3.pkgs.nose # Needed for testing
```
You can move this to `checkInputs`, which will be added only when `doCheck` is `true`.

jtojnar commented on pkgs/applications/misc/onboard/default.nix:
```diff
@@ -0,0 +1,144 @@
+    xorg.libxkbfile
+    libxkbcommon
+    intltool
+    glibcLocales
```
`pkgconfig`, `intltool` and `glibcLocales` are only needed for build itself (not linked or otherwise used in the final product) so they should be moved to `nativeBuildInputs`.

jtojnar commented on pkgs/applications/misc/onboard/default.nix:
```diff
@@ -0,0 +1,144 @@
+  };
+
+  # Uncomment for tests
+  #LC_ALL = "en_US.UTF-8";
```
You can leave this uncommented, it won’t do any harm.

jtojnar reviewed (commented):
this looks very good, just few final details

jtojnar commented on pkgs/applications/misc/onboard/default.nix:
```diff
@@ -0,0 +1,144 @@
+      --replace '"killall",' '"${procps}/bin/pkill", "-x",'
+
+    substituteInPlace  ./Onboard/TextDomain.py \
+      --replace "/bin/bash" "${stdenv.shell}"
```
It looks like /usr/bin/bash is only used as an example of path for test here, no need to patch it.

jtojnar commented on pkgs/applications/misc/onboard/default.nix:
```diff
@@ -0,0 +1,144 @@
+      --replace "/usr/bin/yelp" "${yelp}/bin/yelp"
+
+    substituteInPlace  ./Onboard/SpellChecker.py \
+      --replace "/usr/share/hunspell" ${hunspell}/bin/hunspell \
```
There are also some paths to `aspell`. `(aspellWithDicts (dicts: with dicts; [ en ]))` should be at least added to `checkInputs`.

jtojnar commented on pkgs/applications/misc/onboard/default.nix:
```diff
@@ -0,0 +1,144 @@
+      --replace "/usr/bin/yelp" "${yelp}/bin/yelp"
+
+    substituteInPlace  ./Onboard/SpellChecker.py \
+      --replace "/usr/share/hunspell" ${hunspell}/bin/hunspell \
```
Also, tests require `(hunspellWithDicts [hunspellDicts.en-us])`.

johnramsden replied:
So just put something like the following in ```checkInputs```? 

```
  checkInputs = [
    python3.pkgs.nose
    (aspellWithDicts (dicts: with dicts; [ en ]))
    (hunspellWithDicts [hunspellDicts.en-us])
  ];
```

Do I still just use ```aspell``` and ```hunspell``` as requirements in the function parameter list?  Or will I need to add ```aspellWithDicts``` and ```hunspellWithDicts```.

jtojnar replied:
Yes, for the `nose` and `aspellWithDicts`.

What `buildInputs` does is that it adds the packages to `PATH` and other variables like `NIX_CFLAGS_COMPILE`. Since the source code only contains the absolute path to `hunspell` binary, adding the `hunspellWithDicts` to inputs will not have any effect.

Adding the `aspellWithDicts` to inputs, on the other hand, will be useful, since it is called with just the program name.

Also please note that unless the `PATH` is copied e.g. to a wrapper, or binary in the path fixed during build using, for example, using `which` command – and python packages usually do no such things – it will not have any effect on the resulting package.

----

Yes, you will need to have any derivations you use injected through a function parameter.

johnramsden replied:
So what should I do regarding ```hunspellWithDicts``` then if it will not have any effect?

jtojnar replied:
Replace the `${hunspell}` in the substitution by `${hunspellWithDicts [hunspellDicts.en-us]}`. Or better, define it beforehand:

```nix
let
  hunspell = hunspellWithDicts [hunspellDicts.en-us];
in python3.pkgs.buildPythonApplication rec {
...
}
```

jtojnar reviewed (commented):
I played with the tests a little; about the dconf ones we cannot probably do anything, not sure if it is possible to run daemons during build; the NameErrors are worrying, though.

johnramsden commented:
I'm not sure what changed to start causing this error, but I'm now getting a pkg-config build error: 

> setup.py: pkg-config returned exit code 1
found dconf version 0.26.0
setup.py: sdist needs libgtk-3-dev, libxtst-dev, libxkbfile-dev, libdconf-dev, libcanberra-dev, libhunspell-dev and libudev-dev
builder for ‘/nix/store/ch56jirj3d5y58xbib6z8mmr2xasz2sx-onboard-1.4.1.drv’ failed with exit code 1
cannot build derivation ‘/nix/store/qwjffv4ml4ysc7rd0602g37vlcp8f5rc-system-path.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/m2nc8b31w2s60f99vhmqp7q6cjgjysdg-nixos-system-nixos-17.09.2267.de7eae5d1f3.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/m2nc8b31w2s60f99vhmqp7q6cjgjysdg-nixos-system-nixos-17.09.2267.de7eae5d1f3.drv’ failed

johnramsden commented:
So for whatever reason adding the ```let  hunspell = hunspellWithDicts [hunspellDicts.en-us]; in``` breaks the build causing a pkg-config build error: 

> setup.py: pkg-config returned exit code 1
found dconf version 0.26.0
setup.py: sdist needs libgtk-3-dev, libxtst-dev, libxkbfile-dev, libdconf-dev, libcanberra-dev, libhunspell-dev and libudev-dev
builder for ‘/nix/store/ch56jirj3d5y58xbib6z8mmr2xasz2sx-onboard-1.4.1.drv’ failed with exit code 1
cannot build derivation ‘/nix/store/qwjffv4ml4ysc7rd0602g37vlcp8f5rc-system-path.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/m2nc8b31w2s60f99vhmqp7q6cjgjysdg-nixos-system-nixos-17.09.2267.de7eae5d1f3.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/m2nc8b31w2s60f99vhmqp7q6cjgjysdg-nixos-system-nixos-17.09.2267.de7eae5d1f3.drv’ failed

jtojnar commented:
Okay, then it looks like the hunspell will also need to be in buildInputs after all.

jtojnar commented:
I mean `hunspell.dev`, which is not probably provided by `hunspellWithDicts`.

johnramsden commented:
So I just defined a separate `customHunspell` which contains the dict for the `preBuild`. It seems to be working now. I'm not sure what exactly you meant by using ```hunspell.dev```

jtojnar commented:
Yes, this should be enough. `hunspell` is a [multiple output derivation](https://nixos.org/nixpkgs/manual/#chap-multiple-output) and the files needed for pkgconfig to find it are in the `dev` output. When using `hunspell` as a direct dependency, the dev output is available to onboard, but when it is hidden inside `hunspellWithDicts`, the dev output is not propagated making the pkgconfig unable to find it.

jtojnar commented:
I will test it later but it looks good to merge.

In the future, we might want to look into the remaining test failures, probably in cooperation with upstream. We will also want to allow users to chose which hunspell dictionaries to install.

johnramsden commented:
@jtojnar That sounds like a good idea, adding the ability to choose which dictionaries to use could be customized with install options in the future. 

I've never tried using the dictionary so I'm not sure how they work, but I tried enabling word suggestions and the keyboard would no longer open, so something must be wrong. I have found Upstream to be very responsive to issues in the past so I'm sure if we asked them they would give us a hand.

At least now the core functionality of the keyboard is working though, which is great for accessibility on NixOS. The current alternative xvkbd _works_, but Onboard is a much better keyboard.

@jtojnar Thanks for all your help getting things working properly, your help is much appreciated.

FRidh commented:
> We will also want to allow users to chose which hunspell dictionaries to install.

Typically these kind of things are kept impure for during runtime, that is, you need to explicitly install yourself hunspell with certain dictionaries. The reason is that one otherwise would have to copy interfaces and because its unlikely that you would need multiple versions of hunspell during runtime.

I leave it open whether you should do it like this or not.

jtojnar commented:
> Typically these kind of things are kept impure for during runtime, that is, you need to explicitly install yourself hunspell with certain dictionaries.

In that case we would want to replace `/usr/share/hunspell` with just `hunspell`, add `customHunspell` to `checkInputs` and mention  the need to install them in a comment at the top of the file or in the `longDescription` (not sure which is more idiomatic).

johnramsden commented:
@jtojnar Cgn this what you were thinking?

Also, what is the purpose for replacing `/usr/share/hunspell` with just `hunspell`?

It seems to be working now after enabling the atspi service.

jtojnar commented:
My bad, I misread `/usr/share/hunspell` as `/usr/bin/hunspell` and wanted to change it to just the program name so it would be looked up on `PATH`. Since that is actually a path to the dictionaries, we want to use paths where the dictionaries will be installed to, like `/var/run/current-system/sw/share/hunspell/` and `/home/username/.nix-profile/share/hunspell`. Or we could ignore the `hunspell` class, relying on the `hunspell_cmd` which works seamlessly with `hunspellWithDicts` installed on the system path; though, I did not find out how is the spell-checker supposed to work in onboard or how to switch the backend.

johnramsden commented:
@jtojnar in other words, change `/usr/share/hunspell` to `$out/share/hunspell` then, right?

I couldn't figure out how the spell-checker was supposed to work either.

jtojnar commented:
No, that would not work. `$out` points to the `onboard` directory in the nix store. We would need to use a patch like this, in order to be able to use `hunspell` backend:

```diff
=== modified file 'Onboard/SpellChecker.py'
--- Onboard/SpellChecker.py	2015-02-10 00:09:01 +0000
+++ Onboard/SpellChecker.py	2017-12-05 14:49:31 +0000
@@ -506,6 +506,10 @@
         if dicpath:
             paths.extend(dicpath.split(pathsep))
 
+        datadirs = os.getenv("XDG_DATA_DIRS")
+        if datadirs:
+            paths.extend(map(lambda datadir: os.path.join(datadir, 'hunspell'), datadirs.split(pathsep)))
+
         paths.extend(LIBDIRS)
 
         home = os.getenv("HOME")

```

johnramsden commented:
Okay, I'll apply the patch and set the path to what then?

jtojnar commented:
No need for substitution then. I would also move some things around and add more comments:

```diff
--- a/pkgs/applications/misc/onboard/default.nix
+++ b/pkgs/applications/misc/onboard/default.nix
@@ -28,11 +28,10 @@
 
 let
   customHunspell = hunspellWithDicts [hunspellDicts.en-us];
-in
-python3.pkgs.buildPythonApplication rec {
-  name = "onboard-${version}";
   majorVersion = "1.4";
   version = "${majorVersion}.1";
+in python3.pkgs.buildPythonApplication rec {
+  name = "onboard-${version}";
   src = fetchurl {
     url = "https://launchpad.net/onboard/${majorVersion}/${version}/+download/${name}.tar.gz";
     sha256 = "01cae1ac5b1ef1ab985bd2d2d79ded6fc99ee04b1535cc1bb191e43a231a3865";
@@ -43,8 +42,14 @@
   doCheck = false;
   checkInputs = [
     python3.pkgs.nose
+    # for Onboard.SpellChecker.hunspell_cmd doctests
     customHunspell
+    # for Onboard.SpellChecker.aspell_cmd doctests
     (aspellWithDicts (dicts: with dicts; [ en ]))
+    # for Onboard.SpellChecker.hunspell doctests
+    hunspellDicts.en-us
+    hunspellDicts.es-es
+    hunspellDicts.it-it
   ];
 
   propagatedBuildInputs = [
@@ -83,6 +88,11 @@
     gnome3.dconf
   ];
 
+  patches = [
+    # Allow loading hunspell dictionaries installed in NixOS system path
+    ./use-xdg-datadirs.patch
+  ];
+
   preBuild = ''
     # Unnecessary file, has been removed upstream
     # https://github.com/NixOS/nixpkgs/pull/24986#issuecomment-296114062
@@ -108,7 +118,6 @@
       --replace "/usr/bin/yelp" "${yelp}/bin/yelp"
 
     substituteInPlace  ./Onboard/SpellChecker.py \
-      --replace "/usr/share/hunspell" hunspell \
       --replace "/usr/lib" "$out/lib"
 
     substituteInPlace  ./data/org.onboard.Onboard.service  \
```

jtojnar commented on pkgs/applications/misc/onboard/default.nix:
```diff
@@ -0,0 +1,155 @@
+  meta = {
+    homepage = https://launchpad.net/onboard;
+    description = "An onscreen keyboard useful for tablet PC users and for mobility impaired users.";
+    longDescription = ''
```
I would get rid of the long description. dconf is now added automatically thanks to `propagatedUserEnvPkgs`, changing advanced configuration is not NixOS specific, the a11y bus is not specific to this package… The only thing that matters is hunspell and we are do not even know how to use it.

johnramsden commented:
I went ahead and added the patch, as well as some comments, and did a bit of reorganizing.

jtojnar commented on pkgs/applications/misc/onboard/default.nix:
```diff
@@ -0,0 +1,157 @@
+in python3.pkgs.buildPythonApplication rec {
+  name = "onboard-${version}";
+  majorVersion = "1.4";
+  version = "${majorVersion}.1";
```
It is preferred to move the helper definitions (like version) to the `let` statement in order not to pollute the build environment.

jtojnar commented on pkgs/applications/misc/onboard/default.nix:
```diff
@@ -0,0 +1,157 @@
+  postInstall = ''
+    cp onboard-default-settings.gschema.override.example $out/share/glib-2.0/schemas/10_onboard-default-settings.gschema.override
+
+    ${glib.dev}/bin/glib-compile-schemas $out/share/glib-2.0/schemas/
```
No need for absolute path to `glib-compile-schemas` since it is already on path, thanks to being in build inputs.

jtojnar commented on pkgs/applications/misc/onboard/default.nix:
```diff
@@ -0,0 +1,157 @@
+{ config
```
What is this for?

johnramsden replied:
I I think that remains from when I was using it as one of my own custom packages. I'll remove it.

jtojnar reviewed (approved):
Thanks for your great work, you can squash it now and I will merge it.

jtojnar commented:
The commit message seems to mention uncommenting `checkInputs`.

johnramsden commented:
@jtojnar Where are you referring to?

jtojnar commented:
![image](https://user-images.githubusercontent.com/705123/33642396-1ff09840-da3a-11e7-86ab-c1aa54dc3c56.png)

johnramsden commented:
@jtojnar Whoops, fixed.

jtojnar commented:
Thanks.

johnramsden commented:
@jtojnar Thanks for being so patient with me. I appreciate your help in getting this to work.

jtojnar commented:
Hmm, it should be possible to make the dconf tests work with `dbus-run-session` see

https://github.com/NixOS/nixpkgs/blob/9adbe6476107c16fb5d70430560e57a344613ae4/pkgs/applications/office/paperwork/default.nix#L58-L59

johnramsden commented:
@jtojnar Interesting. How exactly would I go about replacing the check command that's run with `dbus-run-session -- make check`. In the example given dbus is used to run a shell before the check command. Should I just disable checks and run it manually with `dbus-run-session -- make check`?

jtojnar commented:
@johnramsden Simply replacing the `checkPhase` should be sufficient

```diff
--- a/pkgs/applications/misc/onboard/default.nix
+++ b/pkgs/applications/misc/onboard/default.nix
@@ -3,6 +3,7 @@
 , aspellWithDicts
 , at_spi2_core ? null
 , atspiSupport ? true
+, dbus
 , bash
 , glib
 , glibcLocales
@@ -57,6 +59,7 @@ in python3.pkgs.buildPythonApplication rec {
     hunspellDicts.es-es
     hunspellDicts.it-it
 
+    dbus.daemon
     python3.pkgs.nose
   ];
 
@@ -147,6 +151,12 @@ in python3.pkgs.buildPythonApplication rec {
     glib-compile-schemas $out/share/glib-2.0/schemas/
   '';
 
+  checkPhase = ''
+    dbus-run-session \
+      --config-file=${dbus.daemon}/share/dbus-1/session.conf \
+      python setup.py test
+  '';
+
   meta = {
     homepage = https://launchpad.net/onboard;
     description = "An onscreen keyboard useful for tablet PC users and for mobility impaired users.";
```

Though now it seems to hang for me.

jtojnar commented:
I suspect the `AttributeError (module 'pypredict' has no attribute 'lm') ... ERROR` is caused by tests not being wrapped. I tried various combinations of env vars, including the following but nothing really seems to help.

```nix
  checkPhase = ''
    export PYTHONPATH="$out/lib/${python3.libPrefix}/site-packages:$PYTHONPATH"
    export PYDIR=$out/lib/${python3.libPrefix}/site-packages
    export XDG_DATA_DIRS="$out/share:$XDG_DATA_DIRS"
    export GIO_EXTRA_MODULES="$out/lib/gio/modules:$GIO_EXTRA_MODULES"
    dbus-run-session \
      --config-file=${dbus.daemon}/share/dbus-1/session.conf \
      sh -c " \
        ${gnome3.dconf.lib}/libexec/dconf-service & \
        python setup.py test
      "
  '';
```

Maybe some Nix Python people might help too.
